doc/source/analysis.rst

```diff
@@ -1,5 +1,400 @@
+
+>>> m = g.ecount()
+
+To get a sequence of vertices, use their ids and :attr:`Graph.vs`:
+
+>>> for v in g.vs:
+>>>     print(v)
+
+Similarly for edges, use :attr:`Graph.es`:
+
+>>> for e in g.es:
+>>>     print(e)
+
+.. note:: These two attributes are special sequences with their own useful methods. Some are described below, and all can be found in the `API documentation`_.
+
+If you prefer a vanilla edge list, you can use :meth:`Graph.get_edge_list`.
+
+Incidence
+++++++++++++++++++++++++++++++
+To get the vertices at the two ends of an edge, use :attr:`Edge.source` and :attr:`Edge.target`:
+
+>>> e = g.es[0]
+>>> v1, v2 = e.source, e.target
+
+Vice versa, to get the edge if from the source and target vertices, you can use :meth:`Graph.get_eid` or, for multiple pairs of source/targets,
+:meth:`Graph.get_eids`. The boolean version, asking whether two vertices are directly connected, is :meth:`Graph.are_connected`.
+
+To get the edges incident on a vertex, you can use :meth:`Vertex.incident`, :meth:`Vertex.out_edges` and
+:meth:`Vertex.in_edges`. The three are equivalent on undirected graphs but not directed ones, of course:
+
+>>> v = g.vs[0]
+>>> edges = v.incident()
+
+The :meth:`Graph.adjecent` and :meth:`Graph.incident` functions fulfill the same purpose with a slightly different syntax based on vertex ids:
+
+>>> edges = g.incident(0)
+
+To get the full adjacency/incidence list representation of the graph, use :meth:`Graph.get_adjlist`, :meth:`Graph.g.get_inclist()` or, for a bipartite graph, :meth:`Graph.get_incidence`.
+
+Neighborhood
++++++++++++++
+To compute the neighbors, successors, and predecessors, the methods :meth:`Graph.neighbors`, :meth:`Graph.successors` and
+:meth:`Graph.predecessors` are available. The three give the same answer in undirected graphs and have a similar dual syntax:
+
+>>> neis = g.vs[0].neighbors()
+>>> neis = g.neighbors(0)
+
+To get the list of vertives within a certain distance of one or more initial nodes, you can use :meth:`Graph.neighborhood`:
+
+>>> g.neighborhood([0, 1], order=2)
+
+and to find the neighborhood size, there is :meth:`Graph.neighborhood_size`.
+
+Degrees
++++++++
+To compute the degree, in-degree, or out-degree of a node, use :meth:`Vertex.degree`, :meth:`Vertex.indegree`, and :meth:`Vertex.outdegree`:
+
+>>> deg = g.vs[0].degree()
+>>> deg = g.degree(0)
+
+To compute the max degree in a list of vertices, use :meth:`Graph.maxdegree`.
+
+:meth:`Graph.knn` computes the average degree of the neighbors.
+
+Adding and removing vertices and edges
+++++++++++++++++++++++++++++++++++++++
+
+To add nodes to a graph, use :meth:`Graph.add_vertex` and :meth:`Graph.add_vertices`:
+
+>>> g.add_vertex()
+>>> g.add_vertices(5)
+
+This changes the graph `g` in place. You can specify the name of the vertices if you wish.
+
+To remove nodes, use :meth:`Graph.delete_vertices`:
+
+>>> g.delete_vertices(None)  # remove all vertices
```

Comment:
Naïve question from someone who doesn't use the Python interface much: How would I figure out from the docs that `g.delete_vertices(None)` deletes *all* vertices?

Reply:
Hmmm, I did not know that `g.delete_vertices(None)` deletes all the vertices -- it was never intended. It is an unintended consequence of the fact that the function takes something-that-can-be-converted-into-an-igraph_vs_t and that `igraphmodule_PyObject_to_vs_t()` converts `None` to "all vertices".

Maybe we should update `delete_vertices()` to throw an error for `None` as I don't like this syntax. I'd rather have `g.clear()` (but we don't have it now, and it would also need to delete all the _graph_ attributes as well).

Reply:
Haha yeah this one was surprising for me too. I'll remove this from here and start a separate PR implementing `g.clear`.

I don't think the usage for deleting all vertices but leaving the graph attributes will be heavy anyway.

Reply:
see also #320

Reply:
> `igraphmodule_PyObject_to_vs_t()` converts `None` to "all vertices".

Yeah, that reads really funny in English: "Which vertices?" "None" Then it takes *all* of them instead of *none* of them.

Reply:
Okay, the reason is as follows. There are many methods like `degree()` that work on the whole graph by default but that _may_ take a vertex ID or a list of vertex IDs to operate on a subset of vertices instead. For these methods, the `vs` argument needs a default value that we then treat as "all vertices". I took a commonly used shortcut and simply used `None` (which is basically `null` or `nil` in Python), that's why `None` is converted to "all vertices" by `igraphmodule_PyObject_to_vs_t()`. An alternative (and probably a cleaner way) would have been to create a singleton object named `ALL` and then make _this_ the default value instead of `None`. We can add this as a low-priority issue and we can deal with it in the CZI grant if we get to the point where we need to clean up the Python API before 1.0. There would also have to be a deprecation phase where we still allow `None` but show a warning to the user and ask him to use `ALL` instead.